### PR TITLE
Fix for Zoom-out bug

### PR DESCRIPTION
**Pull request by abrin**
trying to create stories and content that allow us to identify two bugs we’ve seen as we expose CanvasPanel.

1. For certain landscape images we’re seeing the zoom out calculation logic is still not 100% right.  Adding some debug in `use-generic-atlas-props ` to isolate that area of the code AND the `CanvasWithLandscapeZoom` story which shows that quite well.

2. We’re using the ContentState API to allow for clean transitions to full-screen and back. But, I believe because the viewport is a significantly different ratio that this is not working properly. Added `CanvasWithMultipleContentStates` story to illustrate this a bit
  a. The user experience is also quite awkward with the image bouncing (perhaps we can set the transition to immediate), or expose that as an option?
  b. it feels like the image is also not positioned with the center of the target area in the center of the viewport (this feels like an atlas bug)

### CODE REVIEW

File: packages/canvas-panel/src/hooks/use-generic-atlas-props.ts [hunk 1]
```diff
@@ -293,19 +293,12 @@ export function useGenericAtlasProps<T = Record<never, never>>(props: GenericAtl
...
-      // then you're zoomed out, this could benefit from more thought / consideration as we move forward
-      // to figure out if there's a more elegant way to identify that next zoom out logic
-      Math.abs(rt.target[4] - rt.getZoomedPosition(ZOOM_IN_FACTOR, {})[4]) > 0.2
-    ) {
-      canZoomOut = true;
-    }
+    // rt.target[4] represents the longest side of the image in the space of the world...
+    // rt.getZoomedPosition(ZOOM_IN_FACTOR, {})[4] represents the longest side of the image if we were to zoom out one button press
+    // if they're the same then you can't zoom out
+    const canZoomOut = Math.round(rt.target[4]) < Math.round(rt.getZoomedPosition(ZOOM_IN_FACTOR, {})[4]);
```
Comment (abrin) [inline packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:299]: it's weird... in retrospect this feels kinda obvious... wonder what I was missing last time @danieltbrennan  @bulbil could benefit from your thoughts here and a little testing?

Comment (stephenwf) [reply packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:299]: `rt.target[4]` is the Max Y and `rt.target[3]` is the Max X. Might be different for landscape.

Comment (abrin) [reply packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:299]: interesting. I must have misunderstood the thread / strand logic when I was debugging last time.  That certainly makes things easier as I can use both the X and Y to check the zoom

Comment (stephenwf) [reply packages/canvas-panel/src/hooks/use-generic-atlas-props.ts]: The runtime has the following getters and setters:

- rt.x
- rt.y
- rt.width
- rt.height

Which read the same `target` property.

<details>
<summary>Background on the numbers</summary>



The original idea (5y ago now!) was to pre-calculate every xywh of the tiles in a grid for every layer of IIIF. Since Atlas was originally a _Manifest_ viewer, this resulted in a _lot_ of tiles. 

So some time was spent figuring out the fastest way to quickly filter long lists of xywh co-ordinates and then also rescale/translate them to draw them in the right place based on the viewport. In pseudo code:

```ts
const viewport = { x: 100, y: 200, width: 512, height: 512 };
const zoom = 1;

for (const tile of tiles) {
  if (!isVisible(tile)) continue;
  let tilePosition = translate(tile, -viewport.x, -viewport.y);
  tilePosition = scale(tilePosition, zoom);
  draw(tile, tilePosition);
}
```

So for deep zoom levels, where the might be thousands of tiles that need to be filtered, scaled, translates and drawn it was decided (at the time) to compact the co-ordinates into an [ArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer)

The buffer was a list of 5 numbers grouped together:
```
[
  1, 0, 0, 100, 100, // { x1: 0, y1: 0, x2: 100, y2: 100, visible: true }
  1, 100, 0, 200, 100, // { x1: 100, y1: 0, x2: 100, y2: 100, visible: true }
  ...
]
```
The first number is a flag the tile "is visible" (1/0) and then x1,y1,x2,y2 in order.

Since this was now roughly a 2D matrix, we can process very large lists of numbers using matrix transforms. So we can take the zoom level, the viewport and create a 3x3 matrix describing how to go from the viewport to the tile.

E.g. translate(5, 8) + scale(2) =
```
[
  2, 0, 5,
  0, 2, 8,
  0, 0, 1,
]
```

[Here is a visualisation of this](https://atlas.stephen.wf/docs/dna)

This transform passes down the layers of Atlas too. So if you have a "Canvas" that is 500,500 and then paint an image service that is 5000,5000 then before it filters the points it will apply the `scale(10)`. Similarly, if you are drawing an annotation or composite image, any translations are baked. 

So we can reuse the transform/scale the "viewport" was also 5 numbers: 
```
1, x1, y1, x2, y2
```
And this can be passed down through the layers of atlas components for doing an intersection and then filter of all the points.

In short, it allows us to transform and filter points by only looping through them once and performing one calculation per co-ordinate. The original idea (5y ago) was to use "SharedArrayBuffers" so that WebGL had access to the same position information and offload almost all of this to the GPU. However, it has still been useful for without that end goal - and Atlas can scale up to loads of tiles and annotations without any issue. 

OpenSeadragon didn't need this for rendering, because it was up to the user to keep track of position and scaling things when you draw them into the OSD world. This made drawing things like annotations and multiple images tricky.

</details>

**Comment by abrin** [comment]
@stephenwf there are some failures here with netlify that look like they might be related to a change in Node requirements... 
```
 @atlas-viewer/iiif-image-api@2.2.0: The engine "node" is incompatible with this module. Expected version ">=18.14.1". Got "16.14.2"
```

_not sure where this is configured though_

**Comment by stephenwf** [comment]
Ah I will have a look at that tomorrow, should be a quick fix. There might be an nvmrc file in the repository

**Comment by abrin** [comment]
there is

**Comment by abrin** [comment]
@stephenwf we think this should be ready. I've taken the liberty of cleaning up the CanvasPanel story to be a bit more extensible now that we're adding use-cases more frequently by trying to push more of the logic into a shared class/ instance
